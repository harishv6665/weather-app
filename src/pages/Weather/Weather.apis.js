import axios from "axios";

const API_KEY = "3dcbba7ca4300e233aff5675be8082ab"; //TODO: pick from secrets

const getWeather = async () => {
    const location = "Munich,de";
    const cnt = "40";
    try {
        return await axios.get("https://api.openweathermap.org/data/2.5/forecast", {
            params: {
                q: location,
                APPID: API_KEY,
                cnt,
                units: "metric"
            }}
        );
    } catch (error) {
        console.error(error);
    }
}

export default {
    getWeather,
};
