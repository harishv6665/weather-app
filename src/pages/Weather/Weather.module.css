.content {
    height: calc(100vh - 56px);
    overflow: auto;
    padding: 24px;
}

@media only screen and (min-width: 768px) {
    .content {
        max-width: 1200px;
        margin: auto;
    }
}
