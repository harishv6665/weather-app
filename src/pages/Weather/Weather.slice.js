import {createSlice} from "@reduxjs/toolkit";

const weatherInitialState = {
    isLoading: true,
    error: "",
    cityInfo: {},
    forecastList: [],
    forecastByDate: {},
};

export const weatherSelector = state => ({
    isLoading: state.weather.isLoading,
    cityInfo: state.weather.cityInfo,
    forecastList: state.weather.forecastList,
    forecastByDate: state.weather.forecastByDate,
});

const weatherSlice = createSlice({
    name: "weather",
    initialState: weatherInitialState,
    reducers: {
        fetchWeatherDataSuccess: (state, action) => {
            state.isLoading = false;
            state.cityInfo = action.payload.city;
            state.forecastList = action.payload.forecastList;
            state.forecastByDate = action.payload.forecastByDate;
        },
        setError: (state, action) => {
            state.isLoading = false;
            state.error = action.payload
        }
    }
});

export const { fetchWeatherDataSuccess, setError } = weatherSlice.actions;

export default weatherSlice;
