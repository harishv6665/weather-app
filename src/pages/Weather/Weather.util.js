import temperatureConverter from "../../shared/utils/temperatureConverter";

/**
 * Groups the array list by date
 *
 * @param {Object[]} list - weather info object list
 *
 * @returns {*} - weather info grouped by date as key
 */

export const groupWeatherByDate = (list) => {
    return list.reduce((acc, item) => {
        const date = item.dt_txt.split(" ", 1);
        if(acc[date]) {
            acc[date].push(item);
        } else {
            acc[date] = [item];
        }
        return acc;
    }, {});
}


/**
 * Function returns average of temperature for a given day's weather info
 *
 * @param {Object[]} day - weather info for single day
 *
 * @returns {number} average temperature for day
 */

export const getAverageTemperature = (day) => {
    const sumTemp = day.reduce((acc, item) => {
        acc += item.main.temp;
        return acc;
    }, 0);
    return sumTemp/day.length;
}


/**
 * Weather info list for cards
 *
 * @param {object} data - weather info grouped by date
 *
 * @returns {Object[]} - array of objects with date and average temperature
 */

export const getWeatherList = (data) => {
    const list = [];
    for(let day in data) {
        const info = {
            date: day,
            avg_temp: getAverageTemperature(data[day]),
        }
        list.push(info);
    }
    return list;
}


/**
 * extracts data to render chart
 *
 * @param {Object[]} data - weather info object list
 * @param {boolean} isCelsius
 *
 * @returns {Object[]} - array of objects with time and temperature
 */

export const getChartData = (data, isCelsius) => {
    return data.map(item => {
        return {
            time: item.dt_txt.substr(item.dt_txt.indexOf(" ") + 1, 5),
            temp: temperatureConverter(item.main.temp, isCelsius),
        }
    })
}
