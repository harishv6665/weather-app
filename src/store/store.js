import { configureStore } from "@reduxjs/toolkit";
import weatherSlice from "../pages/Weather/Weather.slice";

export default configureStore({
  reducer: {
    weather: weatherSlice,
  },
});
