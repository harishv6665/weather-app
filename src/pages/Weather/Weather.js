import React, {useEffect, useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {Loader, Error} from "../../shared/components";
import ForecastList from "./components/ForecastList/ForecastList";
import ForecastChart from "./components/ForecastChart/ForecastChart";
import {getChartData} from "./Weather.util";
import {weatherSelector} from "./Weather.slice";
import {fetchWeatherData} from "./Weather.apis";

const Weather = () => {
    const state = useSelector(weatherSelector);
    const dispatch = useDispatch();
    const [isCelsius, setIsCelsius] = useState(false);
    const [activeForecastDate, setActiveForecastDate] = useState(null);

    useEffect(() => {
        setTimeout(() => {
            dispatch(fetchWeatherData());
        }, 1300)
    }, []);

    if (state.error) {
        return <Error message={state.error} />
    }

    if (state.isLoading) {
        return <Loader />
    }

    return (
        <React.Fragment>
            {!!state.forecastList.length && <ForecastList
                list={state.forecastList}
                isCelsius={isCelsius}
                city={state.cityInfo.name}
                setIsCelsius={() => setIsCelsius(!isCelsius)}
                activeForecastDate={activeForecastDate}
                setActiveForecastDate={date => setActiveForecastDate(date)}
            />}
            {activeForecastDate && <ForecastChart
                forecastDate={activeForecastDate}
                isCelsius={isCelsius}
                data={getChartData(state.forecastByDate[activeForecastDate], isCelsius)}
            />}
        </React.Fragment>
    );
}

Weather.propTypes = {}

Weather.defaultProps = {}

export default Weather;
