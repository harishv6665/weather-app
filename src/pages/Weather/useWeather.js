import {weatherSelector} from "./Weather.slice";
import {useEffect, useState} from "react";
import {fetchWeatherData} from "./Weather.apis";
import {useDispatch, useSelector} from "react-redux";

const useWeather = () => {
    const state = useSelector(weatherSelector);
    const dispatch = useDispatch();
    const [isCelsius, setIsCelsius] = useState(false);
    const [activeForecastDate, setActiveForecastDate] = useState(null);

    useEffect(() => {
        // intentionally added delay to show loader for 1.3 seconds
        setTimeout(() => {
            dispatch(fetchWeatherData());
        }, 1300)
    }, []);

    return {
        state,
        isCelsius,
        setIsCelsius,
        activeForecastDate,
        setActiveForecastDate
    }
}

export default useWeather;
