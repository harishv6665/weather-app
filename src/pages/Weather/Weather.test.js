import React from "react";
import { shallow, mount } from "enzyme";
import Weather from "./Weather";
import store from "../../store/store";
import {Provider} from "react-redux";

describe("Weather tests", () => {

    it("should be defined", () => {
        expect(Weather).toBeDefined();
    });

    it("should match to snapshot", () => {
        const wrapper = shallow(
            <Provider store={store}>
                <Weather />
            </Provider>);
        expect(wrapper).toMatchSnapshot();
    });

    it("should match the DOM structure", () => {
        const wrapper = mount(
            <Provider store={store}>
                <Weather />
            </Provider>);
        expect(wrapper.find("Provider").length).toBe(1);
        expect(wrapper.find("Weather").length).toBe(1);
        expect(wrapper.find("Loader").length).toBe(1);
    });

});
