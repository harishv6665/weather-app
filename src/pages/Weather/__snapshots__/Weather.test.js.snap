// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather tests should match to snapshot 1`] = `ShallowWrapper {}`;
